target/ppc/translate/fixedpoint-impl.c.inc

```diff
@@ -492,3 +492,32 @@ static bool trans_PEXTD(DisasContext *ctx, arg_X *a)
 #endif
     return true;
 }
+
+static bool do_hash(DisasContext *ctx, arg_X *a, bool priv,
+    void (*helper)(TCGv_ptr, TCGv, TCGv, TCGv))
+{
+    TCGv ea = tcg_temp_new();
+
+    if (!(ctx->insns_flags2 & PPC2_ISA310)) {
+        /* if version is before v3.1, this operation is a nop */
+        return true;
+    }
+
+    if (unlikely(priv && ctx->pr)) {
+        /* if instruction is privileged but the context is in user space */
+        gen_priv_exception(ctx, POWERPC_EXCP_PRIV_OPC);
+        return 0;
+    }
+
+    tcg_gen_addi_tl(ea, cpu_gpr[a->ra], a->rt);
```

Comment:
"If RA=0, the instruction form is invalid."
Also, if `NARROW_MODE(ctx)`, we need a `tcg_gen_ext32u_tl(ea, ea)`.

Reply:
Fixed in [7c4ee64](https://github.com/PPC64/qemu/pull/61/commits/7c4ee6465ca66371ddbc84560d25c224608ad651)